### ISSUE TITLE
Read Tags from my AB Micrologix 1400

### ISSUE THREAD

**Issue by lucacarone** [ornl-epics/etherip#2]
Hi All,
I want to use your library to read tags from my PLC AB Micrologix 1400
whenever I try to read tag "N7:0" my code goes in exeception.

What do you think?

this is my code:

package testplc;

import etherip.types.CIPData;
import etherip.types.CIPData.Type;
import etherip.*;

/**
 *
- @author Luca
  */
  public class TestPLC {
  
  /**
  - @param args the command line arguments
    _/
    public static void main(String[] args) {
     EtherNetIP plc = new EtherNetIP("172.16.1.1", 0);
     try{
        plc.connect();
     System.out.println("\n_\n\* Connected:\n_\n");
     System.out.println(plc);
     System.out.println("\n_\n\* Individual read/write:\n*\n");
    
     CIPData value = plc.readTag("N7:0");
     System.out.println(value);
     }
    
     catch(Exception ex)
     {
         System.out.println(ex.getMessage());
     }
  
  }
  }

And this is my output:
EtherIP address '172.16.1.1', session 0x75c83718
Device vendor=0x1, device_type=0xE, revision=0xA02, serial=0x30ED, name='1766-L32BXB B/10.00'

*
- Individual read/write:
  *

Expected CIP_ReadData_Reply (0xCC), got CM_Unconnected_Send_Reply (0xD2)

**Comment by kasemir** [comment]
I don't have that type of Plc to try.
When your u increase the log level to see the complete protocol dump, maybe there is an error reply from the Plc the provides some hints.

**Comment by lucacarone** [comment]
Hi Kasemir,
I have increased my log.
Do you have any idea about why i can't read my tag?

this is my log after connection:


*
* Individual read/write:
*

gen 15, 2015 12:56:39 PM etherip.protocol.Connection write
Migliore: Protocol Encoding
Encapsulation Header
UINT  command           : SendRRData (0x006F)
UINT  length            : 42
UDINT session           : 0x156484DF
UDINT status            : 0x00000000
USINT context[8]        : 'Funstuff'
UDINT options           : 0x00000000
Send RR Data
UDINT interface handle  : 0
UINT timeout            : 0
UINT count (addr., data): 2
UINT address type       : 0x0 (UCMM)
UINT address length     : 0x00
UINT data type          : 0xB2 (Unconnected Message)
UINT data length        : 26
MR Request
USINT service           : CM_Unconnected_Send (0x52)
USINT path              : Path (2 el) Class(0x20) 0x6 (ConnectionManager), instance(0x24) 1
CM_Unconnected_Send
USINT tick_time         : 10
USINT ticks             : -16
UINT message size       : 12
  \/\/\/ embedded message \/\/\/ (12 bytes)
MR Request
USINT service           : CIP_ReadData (0x4C)
USINT path              : Path Symbol(0x91) 'N7:0/0'
USINT elements          : 1
  /\/\/\ embedded message /\/\/\
USINT path size         : 1 words
USINT reserved          : 0
USINT port 1, slot 0

gen 15, 2015 12:56:39 PM etherip.protocol.Connection write
Ottimale: Data sent (66 bytes):
0000 - 6F 00 2A 00 DF 84 64 15 00 00 00 00 46 75 6E 73 - o.*...d.....Funs
0010 - 74 75 66 66 00 00 00 00 00 00 00 00 00 00 02 00 - tuff............
0020 - 00 00 00 00 B2 00 1A 00 52 02 20 06 24 01 0A F0 - ........R. .$...
0030 - 0C 00 4C 04 91 06 4E 37 3A 30 2F 30 01 00 01 00 - ..L...N7:0/0....
0040 - 01 00                                           - ..              

gen 15, 2015 12:56:39 PM etherip.protocol.Connection read
Ottimale: Data read (44 bytes):
0000 - 6F 00 14 00 DF 84 64 15 00 00 00 00 46 75 6E 73 - o.....d.....Funs
0010 - 74 75 66 66 00 00 00 00 00 00 00 00 00 04 02 00 - tuff............
0020 - 00 00 00 00 B2 00 04 00 D2 00 08 00             - ............    

gen 15, 2015 12:56:39 PM etherip.protocol.Connection read
Migliore: Protocol Decoding
Encapsulation Header
UINT  command           : SendRRData (0x006F)
UINT  length            : 20
UDINT session           : 0x156484DF
UDINT status            : 0x00000000 (OK)
USINT context[8]        : 'Funstuff'
UDINT options           : 0x00000000
Received RR Data
UDINT interface handle  : 0
UINT timeout            : 1024
UINT count (addr., data): 2
UINT address type       : 0x0 (UCMM)
UINT address length     : 0
UINT data type          : 0xB2 (Unconnected Message)
UINT data length        : 4
MR Response
USINT service           : CM_Unconnected_Send_Reply (0xD2)
USINT reserved          : 0x0
USINT status            : 0x8 (Service not supported)
USINT ext. stat. size   : 0x0

Expected CIP_ReadData_Reply (0xCC), got CM_Unconnected_Send_Reply (0xD2)
COSTRUITO CON SUCCESSO (tempo totale: 1 secondo)

**Comment by kasemir** [comment]
The 'unconnected send' is used to invoke the 'chip read data' service, but your device replies with CM_Unconnected_Send_Reply status : 0x8 (Service not supported).

I guess this means that the micrologix series does not support the same commands as the controllogix or compactlogix series. So this driver, as it is currently, will simply not work.

**Comment by kasemir** [comment]
The README mentions which PLCs are supported, and what documentation was consulted to learn how they can be interfaced.

Micrologix is not listed, but if you have details on which commands it supports and can implement that, send me a pull request.

**Comment by guyit** [comment]
Dear developer,
Great piece of code!

I'm looking for a java code that simulates RSWho. In other words I'm looking for an application that broadcast a "CIP-ping" request (similar to what RSWho does) and get a list of all Rockwell PLCs includes its device info (such as device name, type, version etc).
I know this currently does not exist but I wonder if it is something you can developed? 

Thanks

**Comment by jonwolf-ibisa** [comment]
Hi, I used lucacarone's code to test a PLC that I'm working on at the moment, I'd like to know how to use the readTag method.

Here is an attach image of the PLC, is this correct if I want to read the Tag Product Name (attribute 7)?

**CIPData value = plc.readTag("Product Name");**

I'm getting this output when running the class TestPLC:

EtherIP address '181.49.9.154', session 0x10022000
Device vendor=0x1, device_type=0xC, revision=0x105, serial=0x6627, name='1756-ENBT/A'

Individual read/write:
null

As you can see I'm getting all the values just fine when printing the whole PLC but I'm getting null when trying to read a single Tag, I think that the PLC do not support getting Single attributes.
How do I know the correct parameter to readTag for each attribute? looking at the image I'm guessing it should be Vendor ID, Device Type, Product Code, Revision, Status, Serial Number, Product Name, State **but all I'm getting are null values**.

Thanks in advance

![plc_molex](https://cloud.githubusercontent.com/assets/17366073/13799527/5b7d8e6a-eaf2-11e5-9fdd-c27639b25fbd.jpg)

**Comment by kasemir** [comment]
As described in the README and in a little more detail in the paper "Interfacing the ControlLogix PLC over Ethernet/IP", http://accelconf.web.cern.ch/AccelConf/ica01/papers/THDT002.pdf, the code is based on

(1) The generic EtherNet/IP specification which defines the basic CIP commands for reading the device info like vendor code.

(2) The Allen Bradley "Logix5000 Data Access" document which defines the CIP service codes
   specific to the ControlLogix 5000 series, for reading and writing tags on the controller.

When the code 'connects' to a PLC, it uses the generic commands from (1) to query the attributes of the 'Identity' object. From that you get
```
EtherIP address '181.49.9.154', session 0x10022000
Device vendor=0x1, device_type=0xC, revision=0x105, serial=0x6627, name='1756-ENBT/A'
```
The 'name' in here is the "Product Name" (attribute code 7) of the Identity Object (object code 1).

That doesn't mean you can read the "ProductName" as a tag.
Tags are not attributes of control net objects.

The AllenBradley Control Logix PLCs support tags.
You define program tags like this in the RSLogix programming software:

![rslogix5000-create-new-tag](https://cloud.githubusercontent.com/assets/1932421/13813006/7c22111e-eb55-11e5-801a-cd259b412a85.png)

https://github.com/EPICSTools/ether_ip/blob/master/testether_ipApp/RSLogix/test.L5K is an example for the *.L5K text version of ladder logic with tags.

The protocol defined in (2) is used to read/write tags.


Bottom line is that this driver is for ControlLogix PLCs. It reads/writes their tags.
If you don't have a  ControlLogix but some other PLC that also uses EtherNet/IP, you can use the basic code for reading attributes of control net objects, but that's just version and vendor code.
You'll have to figure out how your type of PLC provides access to the information that actually matters, i.e. the equivalent of the ControlLogix 'tags'.

**Comment by vusivus** [comment]
Hello I'm also looking to use your library to access data in a Micrologix PLC. I'm actually an intern and my client wants a simple temperature monitoring system and can't afford RSView so has asked me to design a simple system with a simple PLC Like the Micrologix series.

I was about to write my own library using the Logix5000 Data Access document. At the end of the document it clearly states that you can use PCCC commands for different types of PLCs like the SLC series which has the same data structure as the Micrologix series.Here is a snippet from the document:
_
CIP messaging is the preferred method of interacting with Logix5000
controllers, but PCCC messaging is serviceable for many applications, especially
where the legacy communications product is not able to be modified, and where
the Logix5000 customer is willing to do the extra configuration of data table
mappings in the Logix5000 controller. Remote applications that depend on serial
communication over a modem or serial radio link can also use this method.
PCCC commands can also arrive at the controller in the following ways:
• Via the RS-232 serial port
• Encapsulated inside a ControlNet message
• Encapsulated inside a EtherNet/IP message_

And there are PCCC commands for older PLCs like the SLC. As a newbie to the Allen Bradley PLC series, and as someone who hasn't read the Logix5000 Data Access document entirely. Is it possible to somewhat sneak in the PCCC commands via your library, I'm avoiding writing an entire library as this will take time

**Comment by kasemir** [comment]
Well, this piece of software is for Control Logix with Allen-Bradley specific extensions for the Logix5000.

It wouldn't make sense to extend it to serial lines and SLC commands. That's overall very different, it'd be akin to adding a completely different protocol like for Siemens PLCs. Better keep separate protocols in separate libraries.

PCCC commands encapsulated in EtherNet/IP messages is somewhat similar to Logic5000 commands  encapsulated in EtherNet/IP messages.
If you can use pieces of this software to transfer PCCC commands, resulting in your own library: Great.
If the result is close enough to the existing software and you want to submit a pull request: Great.

**Comment by kasemir** [comment]
closed due to inactivity
